argocd/schema_application.go

```diff
@@ -1184,6 +1561,15 @@ func resourceArgoCDApplicationV1() *schema.Resource {
 	}
 }
 
+func resourceArgoCDApplicationV2() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"metadata": metadataSchema("appprojects.argoproj.io"),
+			"spec":     applicationSpecSchemaV2(),
+		},
+	}
+}
+
```

Comment:
It's my understanding that the Application schema does not change, but now the `metadata.namespace` field can be set to a value other than `argocd` therefore, why not keep using `applicationSpecSchemaV1()` ? Unless I'm missing something !

```suggestion
func resourceArgoCDApplicationV2() *schema.Resource {
	return &schema.Resource{
		Schema: map[string]*schema.Schema{
			"metadata": metadataSchema("appprojects.argoproj.io"),
			"spec":     applicationSpecSchemaV1(),
		},
	}
}
```

Reply:
Yes, that is my understanding as well. It is only the AppProject schema that changes (with the addition of the optional field, "source_namespaces").

Reply:
You are correct that the schema itself has not changed in this PR. However, we do need to do a [state migration](https://github.com/oboukili/terraform-provider-argocd/blob/24699dfe067dd4116512c1b7bb6f7158ffbb476d/argocd/resource_argocd_application.go#L56-L60) to ensure that the `Id` on existing resources is updated to include the namespace. For this to happen, my understanding is that we need to upgrade the [schema version](https://github.com/oboukili/terraform-provider-argocd/blob/24699dfe067dd4116512c1b7bb6f7158ffbb476d/argocd/resource_argocd_application.go#L44)  on the application resource. The application resource currently makes uses [applicationSpecSchemaV2](https://github.com/oboukili/terraform-provider-argocd/blob/24699dfe067dd4116512c1b7bb6f7158ffbb476d/argocd/resource_argocd_application.go#L30) (which dropped support for `ksonnet` in the spec schema (V1). So, in this case I believe the code I have implemented is correct. 

That said, I do wonder how important alignment of version numbers between the schema for the `spec` field of the resource and the schema of the resource itself is? Is it easier for someone reading the code if the schema of the `spec` field

https://github.com/oboukili/terraform-provider-argocd/blob/24699dfe067dd4116512c1b7bb6f7158ffbb476d/argocd/resource_argocd_application.go#L30

aligns with the schema of the resource itself

https://github.com/oboukili/terraform-provider-argocd/blob/24699dfe067dd4116512c1b7bb6f7158ffbb476d/argocd/resource_argocd_application.go#L44

My initial thought was to ensure that this aligned (hence the initial addition of `applicationSpecSchemaV3`). However, this isn't strictly necessary since there are no changes to the `spec` field. So, now we are in a situation where we are using `V2` of the `spec` schema in `V3` of the resource schema. Are you good with this or would you prefer if everything aligned? (Hopefully I've explained this well enough :crossed_fingers:)

Reply:
I understand. How about making a little simplification then?
```go
func applicationSpecSchemaV2() *schema.Schema {
  # To support project source namespaces, we do need to do a state migration to ensure that the Id on existing resources is updated to include the namespace. 
  # For this to happen, we need to trigger a dummy schema version upgrade on the application resource.
  return applicationSpecSchemaV1()
}
```